drivers/hwspinlock/sqn_hwspinlock.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2023 Sequans Communications
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sqn_hwspinlock
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/hwspinlock.h>
+
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sqn_hwspinlock);
+
+struct sqn_hwspinlock_data {
+	DEVICE_MMIO_RAM;
+	uint32_t num_locks;
+};
+
+struct sqn_hwspinlock_config {
+	DEVICE_MMIO_ROM;
+	const struct device *hwspinlocks;
+};
+
+static inline mem_addr_t get_lock_addr(const struct device *dev, uint32_t id)
+{
+	return (mem_addr_t)(DEVICE_MMIO_GET(dev) + id * sizeof(uint32_t));
+}
+
+/*
+ * To define CPU id, we use the affinity2 and affinity1
+ * fields of the MIPID.
+ */
+static uint8_t mpidr_to_cpuid(uint64_t mpidr_val)
+{
+	uint8_t cpuid = ((mpidr_val >> 8) & 0x0F) | ((mpidr_val >> 12) & 0xF0);
```

Comment:
Isn't the shift for `Aff2` `16` instead of `12`?

Reply:
What I'm trying to do here is get the first 4 bits of aff1 and put them in the first 4 bits of cpuid and get the first 4 bits of aff2 and put them in the last 4 bits of cpuid. This is why we only offset by 12 instead of 16 and mask with 0xF0.